modules/tensor_mechanics/src/materials/CappedMohrCoulombCosseratStressUpdate.C

```diff
@@ -0,0 +1,120 @@
+  _poissons_ratio = _host_poisson;
+}
+
+void
+CappedMohrCoulombCosseratStressUpdate::setEffectiveElasticity(const RankFourTensor & /*Eijkl*/)
+{
+  _Eij[0][0] = _Eij[1][1] = _Eij[2][2] = _host_E0000;
+  _Eij[0][1] = _Eij[1][0] = _Eij[0][2] = _Eij[2][0] = _Eij[1][2] = _Eij[2][1] = _host_E0011;
+  _En = _Eij[2][2];
+  const Real denom = _Eij[0][0] * (_Eij[0][0] + _Eij[0][1]) - 2 * Utility::pow<2>(_Eij[0][1]);
+  for (unsigned a = 0; a < _num_sp; ++a)
+  {
+    _Cij[a][a] = (_Eij[0][0] + _Eij[0][1]) / denom;
+    for (unsigned b = 0; b < a; ++b)
+      _Cij[a][b] = _Cij[b][a] = -_Eij[0][1] / denom;
+  }
+}
+
+void
+CappedMohrCoulombCosseratStressUpdate::setStressAfterReturnV(
+    const RankTwoTensor & stress_trial,
+    const std::vector<Real> & stress_params,
+    Real /*gaE*/,
+    const std::vector<Real> & /*intnl*/,
+    const yieldAndFlow & /*smoothed_q*/,
+    const RankFourTensor & /*Eijkl*/,
+    RankTwoTensor & stress) const
+{
+  // form the diagonal stress
+  stress = RankTwoTensor(stress_params[0], stress_params[1], stress_params[2], 0.0, 0.0, 0.0);
+  // rotate to the original frame, to give the symmetric part of the stress
+  stress = _eigvecs * stress * (_eigvecs.transpose());
+  // add the non-symmetric parts
+  stress += 0.5 * (stress_trial - stress_trial.transpose());
+}
+
+void
+CappedMohrCoulombCosseratStressUpdate::consistentTangentOperatorV(
+    const RankTwoTensor & stress_trial,
+    const std::vector<Real> & trial_stress_params,
+    const RankTwoTensor & stress,
+    const std::vector<Real> & stress_params,
+    Real gaE,
+    const yieldAndFlow & smoothed_q,
+    const RankFourTensor & elasticity_tensor,
+    bool compute_full_tangent_operator,
+    const std::vector<std::vector<Real>> & dvar_dtrial,
+    RankFourTensor & cto)
+{
+  CappedMohrCoulombStressUpdate::consistentTangentOperatorV(stress_trial,
+                                                            trial_stress_params,
+                                                            stress,
+                                                            stress_params,
+                                                            gaE,
+                                                            smoothed_q,
+                                                            elasticity_tensor,
+                                                            compute_full_tangent_operator,
+                                                            dvar_dtrial,
+                                                            cto);
+
+  if (!compute_full_tangent_operator)
+    return;
+
+  RankFourTensor anti;
+  for (unsigned i = 0; i < _tensor_dimensionality; ++i)
+    for (unsigned j = 0; j < _tensor_dimensionality; ++j)
+      for (unsigned k = 0; k < _tensor_dimensionality; ++k)
+        for (unsigned l = 0; l < _tensor_dimensionality; ++l)
+          anti(i, j, k, l) = 0.5 * (elasticity_tensor(i, j, k, l) - elasticity_tensor(j, i, k, l));
```

Comment:
Is this correction for the nonsymmetric Cosserat formulation?  If yes, or even more strongly if no, would you add a line or two of documentation to explain the use of the Rank4 `anti` please?